algorithms/src/Kokkos_Random.hpp

```diff
@@ -903,24 +905,31 @@ class Random_XorShift64_Pool {
 #else
   Random_XorShift64_Pool()   = default;
 #endif
-  Random_XorShift64_Pool(uint64_t seed) {
-    num_states_ = 0;
 
-    init(seed, execution_space().concurrency());
+  Random_XorShift64_Pool(uint64_t seed)
+      : Random_XorShift64_Pool(execution_space(), seed,
+                               Impl::RandomPoolConstructorIsAsync::no) {}
+
+  Random_XorShift64_Pool(execution_space const& exec, uint64_t seed,
+                         Impl::RandomPoolConstructorIsAsync is_async =
+                             Impl::RandomPoolConstructorIsAsync::yes)
+      : num_states_(exec.concurrency()) {
+    init(exec, seed);
+    if (!static_cast<bool>(is_async)) exec.fence();
   }
 
-  void init(uint64_t seed, int num_states) {
+  void init(execution_space const& exec, uint64_t seed) {
```

Comment:
this is a change of the public interface, right?

Reply:
It is, but I don't think these functions are actually intended to be part of the public interface.

Reply:
Actually they are mentioned in the documentation, but I couldn't find any uses in Trilinos and I don't think the unit tests call them directly either.

Reply:
I recommend we allow that change and we deal with init in #8082 afterwards

Reply:
https://github.com/kokkos/kokkos/pull/5716 and https://github.com/kokkos/kokkos/issues/5714 discussed problems with a default-constructed object. IMHO, we should retain the overload without and execution space instance before discussing deprecation of `init` elsewhere.